File: web/src/wire-manifest.ts
// Generator for the published, language-neutral wire-compatibility manifest
// (`web/wire-compatibility.json`).
//
// Why it exists: the wire revision and directional floor are published as
// TypeScript constants (wire-compatibility.ts) and as Go constants
// (terminal.WireProtocolVersion / MinSupportedClientWireVersion). A consumer
// that is neither — a Dockerfile, a shell release gate, a Python CI script —
// had no machine-readable surface to read them from and resorted to scraping
// the TypeScript source with `sed`, which breaks silently on any reformat.
//
// This module renders that surface FROM the existing WIRE_COMPATIBILITY export,
// so the numbers keep exactly one source of truth in the TypeScript half. The
// rendered text is checked into the repo (see the drift guard in
// wire-manifest.test.ts) because the publish pipeline runs no build step: npm
// and JSR publish the working tree as-is, so a manifest generated at publish
// time would never exist in the tarball.
//
// The manifest deliberately does NOT carry the package version. `npm pkg set
// version` injects that at publish time, after checkout, with no regeneration
// step — an embedded version would be frozen at the repo's placeholder (0.1.0)
// in every published artifact. Consumers read the package version from
// package.json / jsr.json, which the pipeline does rewrite.
//
// Public-artifact contract: see the "Wire compatibility manifest" section of
// web/README.md for what a consumer may rely on and what counts as a breaking
// change.

import { WIRE_COMPATIBILITY, type WireCompatibility } from "./wire-compatibility.js";

/**
 * Schema revision of the manifest FILE, independent of the wire protocol
 * revision it reports. A consumer must reject a `schemaVersion` it does not
 * understand rather than guess at the shape. Bumped only for a breaking
 * change to the manifest's own layout (a removed or repurposed field).
 */
export const WIRE_MANIFEST_SCHEMA_VERSION = 1;

/** Path of the checked-in artifact this module renders, relative to `web/`. */
export const WIRE_MANIFEST_PATH = "wire-compatibility.json";

/** Shape of the generated wire-compatibility manifest. */
export interface WireManifest {
  /** Schema revision of this file's layout (see WIRE_MANIFEST_SCHEMA_VERSION). */
  readonly schemaVersion: number;
  /** Informational provenance note; not part of the parsed contract. */
  readonly generatedBy: string;
  /** The same values the TypeScript `WIRE_COMPATIBILITY` export carries. */
  readonly wireCompatibility: WireCompatibility;
}

/**
 * Builds the manifest object from the live TypeScript constants. Nesting the
 * protocol values under `wireCompatibility` keeps file-level metadata
 * (`schemaVersion`) in a separate namespace from the payload, so a future
 * metadata field can never collide with a protocol field name.
 */
export function buildWireManifest(): WireManifest {
  return {
    schemaVersion: WIRE_MANIFEST_SCHEMA_VERSION,
    generatedBy: "web/src/wire-manifest.ts",
    wireCompatibility: {
      protocolVersion: WIRE_COMPATIBILITY.protocolVersion,
      minimumServerProtocolVersion: WIRE_COMPATIBILITY.minimumServerProtocolVersion,
      incompatibleCloseCode: WIRE_COMPATIBILITY.incompatibleCloseCode,
    },
  };
}

/**
 * Renders the exact bytes of the checked-in manifest: two-space indentation and
 * a trailing newline, which is what `prettier --check` requires of a JSON file
 * in this repo. Comparing rendered text (not parsed objects) is what makes the
 * drift guard able to fail on a hand edit that happens to parse the same.
 */
export function renderWireManifest(): string {
  return `${JSON.stringify(buildWireManifest(), null, 2)}\n`;
}

File: terminal/wiremanifest.go
package terminal

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
)

// The wire-compatibility manifest is this engine's own published data format:
// web/wire-compatibility.json, rendered from the TypeScript constants by
// web/src/wire-manifest.ts and checked in, so it ships inside every npm and JSR
// release. It exists because a consumer that is neither TypeScript nor Go — a
// Dockerfile, a shell release gate — had no machine-readable surface for the
// wire revisions and resorted to scraping the TypeScript source with sed, which
// breaks on any reformat.
//
// A GO consumer needs it for a less obvious reason, and that is why this decoder
// exists. The Go module carries the SERVER half as constants, but a release gate
// compares server against CLIENT, and the client half ships in the npm package —
// pinned independently of the Go module, by design. So the gate has to read the
// vendored artifact's manifest to learn the half its own binary cannot know.
//
// All three family consumers were hand-declaring this struct to do that (and two
// were still scraping TypeScript instead). The schema is the engine's, so the
// decoder is too: a format's producer exporting its own reader beats N consumers
// mirroring it, and this engine already maintained the mirror in a test.
// Remediation, exit codes and refusal wording stay with each gate, which is the
// documented split — the engine says what the numbers are, the consumer says
// which pin to move.

// WireManifestSchemaVersion is the only manifest schema this decoder reads. The
// manifest declares its own version first precisely so a consumer refuses an
// unknown one instead of guessing at a moved field.
//
// It is bumped only for a breaking change to the manifest's own LAYOUT (a removed
// or repurposed field), never for a wire-protocol revision. Its TypeScript twin
// is WIRE_MANIFEST_SCHEMA_VERSION; the in-repo conformance test pins the two
// together.
const WireManifestSchemaVersion = 1

// ErrWireManifestSchema reports a manifest whose schemaVersion this decoder does
// not read.
//
// Worth distinguishing from every other read failure because the REMEDY differs
// and consumers act on it: an unknown schema means the reader is behind the
// artifact, so the fix is to bump the consumer, never to move a wire pin. A gate
// that reported it as a wire incompatibility would send a maintainer to change
// exactly the wrong thing.
var ErrWireManifestSchema = errors.New("terminal: unsupported wire-compatibility manifest schema")

// WireManifest is the decoded manifest.
//
// Flattened relative to the file, deliberately: the nesting under
// `wireCompatibility` exists so file-level metadata can never collide with a
// protocol field name, which is a property of the FILE's layout and not of the
// values a caller wants.
type WireManifest struct {
	// SchemaVersion is the layout revision, already validated against
	// WireManifestSchemaVersion by the time a caller sees it.
	SchemaVersion int
	// ProtocolVersion is the client's current wire revision.
	ProtocolVersion int
	// MinimumServerProtocolVersion is the client's directional floor: the oldest
	// server revision it will still decode.
	MinimumServerProtocolVersion int
	// IncompatibleCloseCode is the close code the runtime handshake uses to
	// refuse a declared-incompatible peer. Informational for a build gate.
	IncompatibleCloseCode int
}

// ReadWireManifest reads and validates the wire-compatibility manifest at path,
// which is the `wire-compatibility.json` at the root of a vendored engine npm
// package.
//
// Every failure is an error rather than a zero value, and none of them is a
// compatibility verdict: an unreadable, malformed, unknown-schema or
// non-positive manifest means the caller cannot SEE the client's declaration,
// which is a broken gate. A caller must not turn any of these into "bump a pin".
//
// The non-positive check belongs here rather than in the comparator because it
// is a property of the file: WirePairIncompatibility treats a non-positive input
// as a caller error, and a manifest that carries one is malformed at the source.
func ReadWireManifest(path string) (WireManifest, error) {
	data, err := os.ReadFile(path) // #nosec G304 -- the path is the caller's own build argument
	if err != nil {
		return WireManifest{}, fmt.Errorf("terminal: read wire-compatibility manifest: %w", err)
	}
	return DecodeWireManifest(data)
}

// DecodeWireManifest is ReadWireManifest over bytes already in hand, for a caller
// that obtained the artifact some other way (an embedded copy, a registry fetch).
func DecodeWireManifest(data []byte) (WireManifest, error) {
	// Only the published fields are declared; an unknown field is tolerated so an
	// additive manifest revision does not have to break every consumer at once,
	// which is what the schemaVersion gate is for.
	var raw struct {
		WireCompatibility struct {
			ProtocolVersion              int `json:"protocolVersion"`
			MinimumServerProtocolVersion int `json:"minimumServerProtocolVersion"`
			IncompatibleCloseCode        int `json:"incompatibleCloseCode"`
		} `json:"wireCompatibility"`
		SchemaVersion int `json:"schemaVersion"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		return WireManifest{}, fmt.Errorf("terminal: parse wire-compatibility manifest: %w", err)
	}
	if raw.SchemaVersion != WireManifestSchemaVersion {
		return WireManifest{}, fmt.Errorf("%w: manifest declares %d, this build reads %d",
			ErrWireManifestSchema, raw.SchemaVersion, WireManifestSchemaVersion)
	}
	m := WireManifest{
		SchemaVersion:                raw.SchemaVersion,
		ProtocolVersion:              raw.WireCompatibility.ProtocolVersion,
		MinimumServerProtocolVersion: raw.WireCompatibility.MinimumServerProtocolVersion,
		IncompatibleCloseCode:        raw.WireCompatibility.IncompatibleCloseCode,
	}
	if m.ProtocolVersion <= 0 || m.MinimumServerProtocolVersion <= 0 {
		return WireManifest{}, fmt.Errorf(
			"terminal: wire-compatibility manifest carries no usable revisions (protocolVersion %d, minimumServerProtocolVersion %d)",
			m.ProtocolVersion, m.MinimumServerProtocolVersion)
	}
	return m, nil
}
